internal/resources/kustomization/scope/scope.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright © 2023 VMware, Inc. All Rights Reserved.
+SPDX-License-Identifier: MPL-2.0
+*/
+
+package scope
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"golang.org/x/exp/slices"
+
+	kustomizationclustermodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/kustomization/cluster"
+	kustomizationclustergroupmodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/kustomization/clustergroup"
+	commonscope "github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/common/scope"
+)
+
+// ScopedFullname is a struct for all types of kustomization full names.
+type ScopedFullname struct {
+	Scope                commonscope.Scope
+	FullnameCluster      *kustomizationclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceFluxcdKustomizationFullName
+	FullnameClusterGroup *kustomizationclustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupNamespaceFluxcdKustomizationFullName
+}
+
+var (
+	ScopesAllowed = [...]string{commonscope.ClusterKey, commonscope.ClusterGroupKey}
+	ScopeSchema   = commonscope.GetScopeSchema(
+		commonscope.WithDescription(fmt.Sprintf("Scope for the kustomization, having one of the valid scopes: %v.", strings.Join(ScopesAllowed[:], `, `))),
+		commonscope.WithScopes(ScopesAllowed[:]))
+)
+
+func ConstructScope(d *schema.ResourceData, name, namespace, orgID string) (scopedFullnameData *ScopedFullname) {
+	value, ok := d.GetOk(commonscope.ScopeKey)
+
+	if !ok {
+		return scopedFullnameData
+	}
+
+	data, _ := value.([]interface{})
+
+	if len(data) == 0 || data[0] == nil {
+		return scopedFullnameData
+	}
+
+	scopeData := data[0].(map[string]interface{})
```

Comment:
possible panic

Reply:
We have a length check for the data and a nil check for the first element right before this type assertion.

```
if len(data) == 0 || data[0] == nil {
	return scopedFullnameData
}
```

I have manually tested and verified the scope block multiple times. I did not encounter a panic yet.


Also, we have been following this in every other scope implementation -
[resources/policy/scope/scope_schema.go#L65](https://github.com/vmware/terraform-provider-tanzu-mission-control/blob/143c1e710f639065d1924cfe57d479be64720867/internal/resources/policy/scope/scope_schema.go#L65)